kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/RecordEncryptionFilterIT.java

```diff
@@ -218,6 +226,65 @@ void roundTripManyRecordsFromDifferentProducers(KafkaCluster cluster, Topic topi
+                .withConfig("kms", testKmsFacade.getKmsServiceClass().getSimpleName())
+                .withConfig("kmsConfig", testKmsFacade.getKmsServiceConfig())
+                .withConfig("experimental", Map.of("encryptionDekRefreshAfterWriteSeconds", edekExpiry.toSeconds()))
+                .withConfig("selector", TemplateKekSelector.class.getSimpleName())
+                .withConfig("selectorConfig", Map.of("template", TEMPLATE_KEK_SELECTOR_PATTERN))
+                .build());
+
+        var messageBeforeKeyRotation = "hello world, old key";
+        var messageAfterKeyRotation = "hello world, new key";
+        try (var tester = kroxyliciousTester(builder);
+                var producer = tester.producer()) {
+            producer.send(new ProducerRecord<>(topic.name(), messageBeforeKeyRotation)).get(5, TimeUnit.SECONDS);
+
+            Thread.sleep(edekExpiry.toMillis());
+            // refreshAfterWrite is asynchronous, so it is likely the first message triggers the refresh but uses the existing cached EDEK
+            producer.send(new ProducerRecord<>(topic.name(), messageAfterKeyRotation)).get(5, TimeUnit.SECONDS);
+            producer.send(new ProducerRecord<>(topic.name(), messageAfterKeyRotation)).get(5, TimeUnit.SECONDS);
+            producer.send(new ProducerRecord<>(topic.name(), messageAfterKeyRotation)).get(5, TimeUnit.SECONDS);
+
+            // check we can decrypt
+            try (var consumer = tester.consumer()) {
+                consumer.subscribe(List.of(topic.name()));
+                var records = consumer.poll(Duration.ofSeconds(2));
+                assertThat(records.iterator())
+                        .toIterable()
+                        .extracting(ConsumerRecord::value)
+                        .containsExactly(messageBeforeKeyRotation, messageAfterKeyRotation, messageAfterKeyRotation, messageAfterKeyRotation);
+            }
+
+            assertMoreThanOneEdekUsed(topic, testKmsFacade, directConsumer);
+        }
+    }
+
+    private static void assertMoreThanOneEdekUsed(Topic topic, TestKmsFacade<?, ?, ?> testKmsFacade, KafkaConsumer<byte[], byte[]> directConsumer) {
+        directConsumer.subscribe(List.of(topic.name()));
+        ConsumerRecords<byte[], byte[]> records = directConsumer.poll(Duration.ofSeconds(2));
+        Set<?> edeks = StreamSupport.stream(records.spliterator(), false).map(record -> {
+            List<byte[]> encryptionVersions = StreamSupport.stream(record.headers().headers("kroxylicious.io/encryption").spliterator(), false)
+                    .map(Header::value).toList();
+            assertThat(encryptionVersions).hasSize(1).singleElement(BYTE_ARRAY).hasSize(1);
+            Encryption encryption = EncryptionResolver.ALL.fromSerializedId(encryptionVersions.getFirst()[0]);
+            return encryption.wrapper().readSpecAndEdek(ByteBuffer.wrap(record.value()), testKmsFacade.getEdekSerde(), (cipherManager, o) -> o);
```

Comment:
It turns out we do have code available to access the edek, though obtaining the Edek Serde is a little awkward, I made the TestFacade aware of which EdekSerde is associated with them but really that's something hidden in the KMS implementation that could drift apart.

Reply:
maybe instead of supplying the specific Edek Serde for each implementation we could supply a byte[] serde. We don't really need to decode the EDEK into a specific form.

Reply:
I think its good enough for now. Probably indicative of a problem for implementations other than kroxylicious trying to decode our format (which might not be a real issue)

Reply:
nope, some Serdes just know the lengths of the things they are serding, so it has to differ by KMS. Hmm

Reply:
I'm not sure I follow, if there is potential drift over time in whats persisted in the topic doesn't that represent a problem for non runtime clients? I mean if its a problem for our tests its a problem for other users, no?

Reply:
I think we can actually make a byte[] serde for deserializing. The Serde is passed a slice buffer that contains only the Edek bytes.